include/unicode.h

```diff
@@ -0,0 +1,90 @@
+/*
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ *  Copyright (C) 2024-2024  The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#ifndef DOSBOX_UNICODE_H
+#define DOSBOX_UNICODE_H
+
+#include <string>
+
+// Get recommended DOS code page to render the UTF-8 strings to. This
+// might not be the code page set using KEYB command, for example due
+// to emulated hardware limitations, or duplicated code page numbers
+uint16_t get_utf8_code_page();
+
+// Specifies what to do if the DOS code page does not contain character
+// representing given Unicode grapheme
+enum class UnicodeFallback {
+
+	// Convert all unknown graphemes to 0 - to be used in code like TREE
+	// command implementation, which has it's own specialized ASCII fallback
+	// drawing code
+	Null,
```

Comment:
It took me a while to figure out what was going on with the searching for `\0` in the returned string.  As far as I can tell, in all use-cases that use this mode, the returned string is never used if it failed to convert.  Instead, they're checking if the returned string has null terminators in the middle of it and, if so, treats it as an error condition.

I would suggest instead returning a `std::optional<std::string>` or just an empty string to indicate failure.  The other two modes appear to never fail but instead provide a fallback so you could also remove this mode from the enum and put it into a separate function that can fail if that makes more sense.

Example: `std::optional<std::string> utf8_to_dos_or_fail` and leave the other two modes simply returning `std::string`.

I'll leave that up to you to decide if that makes sense or not.

Reply:
This was done with one future use case in mind: the filesystem, namely Unicode file names (UTF-16 on Windows, UTF-8 elsewhere). I thought in case the function returns some `\0`'s (usually when the current code page does not have a suitable character) the filesystem emulation could still handle it somehow, like the host `FOOϪBAR.TXT` file could be presented as `FOOBAR~1.TXT` or `FOO_BA~1.TXT`to the guest software, in a similar way long file names are shown.

So maybe I should keep the `Null`, but add additional mode, for example called `EmptyString`? Right now you are probably the most experienced with filesystem code, so I'm asking you for an advice :)

Reply:
Hmm... maybe that would be useful for the filesystem but it's hard to tell in advance what the best approach is until someone implements it.

The DOS to host file mapping is done in `drive_cache.cpp` which I have not touched.  I took a look at it now and I think it will need a fair bit of refactoring to properly handle Unicode.  Not sure how well it deals with non-ASCII characters in filenames at the moment but I'm going to assume not very well.

Everything is done using C style `char` arrays and pointers and heavy use of `strcpy` `strlen` and friends.  It's getting the host file names with `opendir` and `readdir` which may give back something utf-8 encoded but may blow up on Windows (don't think these functions can ever give back utf-16 encoding).

>  So maybe I should keep the Null, but add additional mode, for example called EmptyString? 

Yeah, that's probably fine.  Maybe in the future the drive cache can make use of a "token" null terminator and replace it with something sensible but we can cross that bridge when we come to it.

Reply:
In general, adding code we don't need right now but to cater for a future scenario is not a good idea It's not easy to predict the future 😄 Unless you're quite sure you'll need something... but even that can often change.

So I'm inclined to say remove it if we don't need it right now.

Reply:
OK, I have changed the fallback mode. It'll be easy to provide the _Null_ one if needed in the future.